log.go

```diff
@@ -2,12 +2,119 @@ package didplc
 
 import (
 	"fmt"
+	"sync"
 	"time"
 
-	"github.com/bluesky-social/indigo/atproto/crypto"
 	"github.com/bluesky-social/indigo/atproto/syntax"
 )
 
+type OpStatus struct {
+	CreatedAt   time.Time // the only immutable field here
+	Nullified   bool
+	LastChild   string // CID
+	AllowedKeys []string
+}
+
+// Note: LogValidationContext is designed such that it could later be turned into an interface,
+// optionally backed by a db rather than in-memory
+// Note: ops are globally unique by CID, so opStatus map can be shared across all DIDs
+type LogValidationContext struct {
+	head     map[string]string    // DID -> CID, tracks most recent valid op for a particular DID
+	opStatus map[string]*OpStatus // CID -> OpStatus
+	lock     sync.RWMutex
+}
+
+func (c *LogValidationContext) GetValidationContext(did string, cid string) (string, *OpStatus, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	head, exists := c.head[did]
+	if !exists {
+		if cid != "" {
+			return "", nil, fmt.Errorf("DID not found")
+		}
+		return head, nil, nil // Not an error condition! just means DID is not created yet
```

Comment:
I think `head` here should be empty string? I guess it implicitly is that if exists is false, but this confused me a bunch on first read through

Reply:
Would it be clearer if I switched it to this?
```go
head := c.head[did]
if head == "" {
    ...
```
I wanted to use the `head` variable rather than a `""` literal in the return statement to make it clear that the current head value is being returned, rather than an error value (even though the head value happens to be `""`).

Reply:
I think re-using a zero val like that just gives me an itch in general. Having it be explicit literal in the return statement itself feels much clearer.